# Homework 8 - Ishaan Sathaye

## Part A: Concepts and Derivations

In the last homework, you created a Neural Network with the following properties:
- Three sets of weights and biases for the initial layer, $\overrightarrow{w_1}$, $\overrightarrow{w_2}$, $\overrightarrow{w_3}$; $c_1$, $c_2$, $c_3$
- One weight set and bias for the hidden layer, $\overrightarrow{v} = {v_1, v_2, v_3}$, $d$
- A Sigmoid activation function,
$$ g(u) = \frac{1}{1 + e^{-u}} $$
- A loss function of squared-error on the predicted probabilities of Class 1, compared to y values of 0 or 1.

### Question 1: Softplus

Consider instead using the Softplus activation function for the first layer only:
$$ g(u) = log(1 + e^u) $$

Derive the back propagation gradient for $v_1$ and for $w_11$ with this activation function.

- first layer: $g(u) = log(1 + e^u)$
- second layer: $g(u) = \frac{1}{1 + e^{-u}}$
- Gradient for $v_1$:
    - $\frac{\partial L}{\partial v_1} = \frac{\partial L}{\partial \hat{p_i}} \frac{\partial \hat{p_i}}{\partial u} \frac{\partial u}{\partial v_1}$
    - $\frac{\partial L}{\partial \hat{p_i}} = \frac{2}{n} (\hat{p_i} - y_i)$
    - $\frac{\partial \hat{p_i}}{\partial u} = \hat{p_i} (1 - \hat{p_i})$
    - $\frac{\partial u}{\partial v_1} = x_{1i}, \text{ where } u = v_1 x_{1i} + v_2 x_{2i} + v_3 x_{3i} + d$
    - $\frac{\partial L}{\partial v_1} = \frac{2}{n} (\hat{p_i} - y_i) \hat{p_i} (1 - \hat{p_i}) x_{1i}$
- Gradient for $w_11$:
    - $\frac{\partial L}{\partial w_{11}} = \frac{\partial L}{\partial \hat{p_i}} \frac{\partial \hat{p_i}}{\partial u} \frac{\partial u}{\partial x_{1i}} \frac{\partial x_{1i}}{\partial z_{1}} \frac{\partial z_{1}}{\partial w_{11}}$
    - $\frac{\partial u}{\partial x_{1i}} = v_1$
    - $\frac{\partial x_{1i}}{\partial z_{1}} = \frac{1}{1 + e^{-z_1}}, \text{ since } x_{1i} = g(z_1)$
    - $\frac{\partial z_{1}}{\partial w_{11}} = x_{1i}$
    - $\frac{\partial L}{\partial w_{11}} = \frac{2}{n} \sum_{i=1}^{n} (\hat{p_i} - y_i) \hat{p_i} (1 - \hat{p_i}) v_1 \frac{1}{1 + e^{-z_1}} x_{1i}$

### Question 2: ReLu

Consider instead using the ReLu activation function for the first layer only:
$$ g(u) = u \text{ if } u \geq 0; 0 \text{ otherwise} $$

Derive the back propagation gradient for $v_1$ and for $w_11$ with this activation function.

- first layer: $g(u) = u \text{ if } u \geq 0; 0 \text{ otherwise}$
- second layer: $g(u) = \frac{1}{1 + e^{-u}}$
- Gradient for $v_1$:
    - same as above
    - $\frac{\partial L}{\partial v_1} = \frac{2}{n} (\hat{p_i} - y_i) \hat{p_i} (1 - \hat{p_i}) x_{1i}$
- Gradient for $w_11$:
    - same as above
    - $x_{1i} = z_1 \text{ if } z_1 \geq 0; 0 \text{ otherwise}$
    - $\frac{\partial x_{1i}}{\partial z_{1}} = 1 \text{ if } z_1 \geq 0; 0 \text{ otherwise}$
    - $\frac{\partial z_{1}}{\partial w_{11}} = x_{1i}$
    - $\frac{\partial L}{\partial w_{11}} = \frac{2}{n} \sum_{i=1}^{n} (\hat{p_i} - y_i) \hat{p_i} (1 - \hat{p_i}) v_1 x_1 \cdot 1 \text{ if } z_1 \geq 0; 0 \text{ otherwise}$

### Question 3: SVC Loss

Consider instead the **loss** function
$$ L = \frac{1}{n} \sum_{i=1}^{n} max(0, 1 - y_i \cdot \hat{y_i}) $$

Derive the back propagation gradient for $v_1$ and for $w_11$ with this loss function, with ReLu activation on the first layer.

- Gradient for $v_1$:
    - $\frac{\partial L}{\partial v_1} = \frac{\partial L}{\partial \hat{p_i}} \frac{\partial \hat{p_i}}{\partial u} \frac{\partial u}{\partial v_1}$
    - $\frac{\partial L}{\partial \hat{p_i}} = \frac{1}{n} \sum_{i=1}^{n} \frac{\partial max(0, 1 - y_i \cdot \hat{y_i})}{\partial \hat{p_i}}$
    - $\frac{\partial \hat{p_i}}{\partial u} = \hat{p_i} (1 - \hat{p_i})$
    - $\frac{\partial u}{\partial v_1} = x_{1i}, \text{ where } u = v_1 x_{1i} + v_2 x_{2i} + v_3 x_{3i} + d$
    - $\frac{\partial L}{\partial v_1} = \frac{-1}{n} \sum_{i=1}^{n} \frac{\partial max(0, 1 - y_i \cdot \hat{y_i})}{\partial \hat{p_i}} \hat{p_i} (1 - \hat{p_i}) x_{1i}$
- Gradient for $w_11$:
    - $\frac{\partial L}{\partial w_{11}} = \frac{\partial L}{\partial \hat{p_i}} \frac{\partial \hat{p_i}}{\partial u} \frac{\partial u}{\partial x_{1i}} \frac{\partial x_{1i}}{\partial z_{1}} \frac{\partial z_{1}}{\partial w_{11}}$
    - $\frac{\partial u}{\partial x_{1i}} = v_1$
    - $\frac{\partial x_{1i}}{\partial z_{1}} = 1 \text{ if } z_1 \geq 0; 0 \text{ otherwise}$
    - $\frac{\partial z_{1}}{\partial w_{11}} = x_{1i}$
    - $\frac{\partial L}{\partial w_{11}} = \frac{-1}{n} \sum_{i=1}^{n} \frac{\partial max(0, 1 - y_i \cdot \hat{y_i})}{\partial \hat{p_i}} \hat{p_i} (1 - \hat{p_i}) v_1 \cdot 1 \text{ if } z_1 \geq 0; 0 \text{ otherwise}$

## Part B: Computer Implementation

### Question 1: NN with Squared-Error Loss

Update your neural network function so that it uses a Sigmoid activation function for the first layer by default, but includes an **optional argument** allowing for either ReLu or SoftPlus activation instead. (Continue to use Sigmoid for the final layer activation.)

Use your function to fit an Indica vs Sativa model on a 80% training set from the cannabis data.

In [32]:
import pandas as pd

cannabis = pd.read_csv('../hw3/cannabis_full.csv')
cannabis_hybrid = cannabis[cannabis['Type'] == 'hybrid']

# remove hybrid strains
cannabis = cannabis[cannabis['Type'] != 'hybrid']

# remove effects and flavors
cannabis = cannabis.drop(columns=['Effects', 'Flavor'])

# map type to 0 or 1
cannabis['Type'] = cannabis['Type'].map({'indica': 0, 'sativa': 1})

# drop null values
cannabis = cannabis.dropna()

from sklearn.linear_model import LogisticRegression

effect_cols = cannabis.columns[cannabis.columns.get_loc("Creative"):cannabis.columns.get_loc("Mouth")+1]
flavors_cols = cannabis.columns[cannabis.columns.get_loc("Earthy"):cannabis.columns.get_loc("Pear")+1]

# Logistic Regression using only effect predictors
m1 = LogisticRegression()
m1.fit(cannabis[effect_cols], cannabis['Type'])

# Logistic Regression using only flavor predictors
m2 = LogisticRegression()
m2.fit(cannabis[flavors_cols], cannabis['Type'])

# Logistic Regression using only rating predictor
m3 = LogisticRegression()
m3.fit(cannabis[['Rating']], cannabis['Type'])

import numpy as np

w1 = np.concatenate([m1.coef_[0], np.zeros(len(flavors_cols) + 1)])
w2 = np.concatenate([np.zeros(len(effect_cols)), m2.coef_[0], np.zeros(1)])
w3 = np.concatenate([np.zeros(len(effect_cols) + len(flavors_cols)), m3.coef_[0]])

c1 = m1.intercept_[0]
c2 = m2.intercept_[0]
c3 = m3.intercept_[0]

In [33]:
def loss(pred, target):
    return (pred - target) ** 2

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_der(x):
    return sigmoid(x) * (1 - sigmoid(x))

def relu(x):
    return max(0, x)

def relu_der(x):
    return 1 if x > 0 else 0

def softplus(x):
    return np.log(1 + np.exp(x))

def softplus_der(x):
    return np.exp(x) / (1 + np.exp(x))

def check_stopping_condition_all(v, d, w1, w2, w3, c1, c2, c3, threshold=1e-4):
    return (np.linalg.norm(v) < threshold and np.abs(d) < threshold and
        np.linalg.norm(w1) < threshold and np.linalg.norm(w2) < threshold
        and np.linalg.norm(w3) < threshold and np.abs(c1) < threshold and
        np.abs(c2) < threshold and np.abs(c3) < threshold)

In [34]:
def gd_perceptron_all(train_pred, train_target, w1, w2, w3, c1, c2, c3, learning_rate=0.01, epochs=500,
                      activation='sigmoid'):
    # initialize v and d
    v = np.array([1/3, 1/3, 1/3])
    d = 0

    # set the activation function
    if activation == 'sigmoid':
        act_func = sigmoid
        act_der = sigmoid_der
    elif activation == 'relu':
        act_func = relu
        act_der = relu_der
    elif activation == 'softplus':
        act_func = softplus
        act_der = softplus_der
    
    for epoch in range(1, epochs+1):
        
        total_loss = 0
        
        for preds, target in zip(train_pred, train_target):
            preds = np.array(preds)
            
            # forward pass
            z1 = np.dot(preds, w1) + c1
            z2 = np.dot(preds, w2) + c2
            z3 = np.dot(preds, w3) + c3
            u1 = act_func(z1)
            u2 = act_func(z2)
            u3 = act_func(z3)
            u = np.array([u1, u2, u3])
            
            # hidden layer
            hidden_layer_input = np.dot(u, v) + d

            output = act_func(hidden_layer_input)
            
            # loss
            current_loss = loss(output, target)
            total_loss += current_loss
            
            # backpropagation
            output_error = output - target
            deriv = act_der(hidden_layer_input)
            
            # gradients for v and d
            grad_v = output_error * deriv * u
            grad_d = output_error * deriv
            
            grad_u = output_error * deriv * v
            grad_z1 = grad_u[0] * act_der(z1)
            grad_z2 = grad_u[1] * act_der(z2)
            grad_z3 = grad_u[2] * act_der(z3)
            grad_w1 = grad_z1 * preds
            grad_w2 = grad_z2 * preds
            grad_w3 = grad_z3 * preds
            grad_c1 = grad_z1
            grad_c2 = grad_z2
            grad_c3 = grad_z3
            
            w1 -= learning_rate * grad_w1
            w2 -= learning_rate * grad_w2
            w3 -= learning_rate * grad_w3
            c1 -= learning_rate * grad_c1
            c2 -= learning_rate * grad_c2
            c3 -= learning_rate * grad_c3
            v -= learning_rate * grad_v
            d -= learning_rate * grad_d

        if check_stopping_condition_all(v, d, w1, w2, w3, c1, c2, c3, 1e-4):
            break

    return w1, w2, w3, c1, c2, c3, v, d

In [35]:
def predict_with_fitted_model(preds, w1, w2, w3, c1, c2, c3, v, d, activation='sigmoid'):
    preds = np.array(preds)
    if activation == 'sigmoid':
        act_func = sigmoid
    elif activation == 'relu':
        act_func = relu
    elif activation == 'softplus':
        act_func = softplus

    z1 = np.dot(preds, w1) + c1
    z2 = np.dot(preds, w2) + c2
    z3 = np.dot(preds, w3) + c3
    u1 = act_func(z1)
    u2 = act_func(z2)
    u3 = act_func(z3)
    u = np.array([u1, u2, u3])
    # hidden layer
    hidden_layer_input = np.dot(u, v) + d
    output = act_func(hidden_layer_input)
    
    return output * 100

In [36]:
# split data into training and testing

def split_data(X, y, test_size=0.2):
    n = X.shape[0]
    indices = list(range(n))
    split = int(test_size * n)
    train_indices = indices[split:]
    test_indices = indices[:split]
    X_train = X[train_indices]
    y_train = y[train_indices]
    X_test = X[test_indices]
    y_test = y[test_indices]
    return X_train, y_train, X_test, y_test

cannabis_preds = cannabis[effect_cols.tolist() + flavors_cols.tolist() + ['Rating']].values
cannabis_target = cannabis['Type'].values
X_train, y_train, X_test, y_test = split_data(cannabis_preds, cannabis_target)

# fit the model
sig_w1, sig_w2, sig_w3, sig_c1, sig_c2, sig_c3, sig_v, sig_d = gd_perceptron_all(X_train, y_train, w1, w2, w3, c1, c2, c3, activation='sigmoid')
relu_w1, relu_w2, relu_w3, relu_c1, relu_c2, relu_c3, relu_v, relu_d = gd_perceptron_all(X_train, y_train, w1, w2, w3, c1, c2, c3, activation='relu')
softplus_w1, softplus_w2, softplus_w3, softplus_c1, softplus_c2, softplus_c3, softplus_v, softplus_d = gd_perceptron_all(X_train, y_train, w1, w2, w3, c1, c2, c3, activation='softplus')

# print values
print('Sigmoid Activation')
print('w1:', sig_w1)
print('w2:', sig_w2)
print('w3:', sig_w3)
print('c1:', sig_c1)
print('c2:', sig_c2)
print('c3:', sig_c3)
print('v:', sig_v)
print('d:', sig_d)
print()
print('ReLU Activation')
print('w1:', relu_w1)
print('w2:', relu_w2)
print('w3:', relu_w3)
print('c1:', relu_c1)
print('c2:', relu_c2)
print('c3:', relu_c3)
print('v:', relu_v)
print('d:', relu_d)
print()
print('Softplus Activation')
print('w1:', softplus_w1)
print('w2:', softplus_w2)
print('w3:', softplus_w3)
print('c1:', softplus_c1)
print('c2:', softplus_c2)
print('c3:', softplus_c3)
print('v:', softplus_v)
print('d:', softplus_d)


Sigmoid Activation
w1: [ 0.075477   -2.12873377 -0.32514469  0.69564431 -0.88976412 -1.01296195
  0.27093805  0.13004627  0.10504993  0.31907568 -0.26513987 -0.22533783
 -1.95733903  0.          0.         -0.02699409  0.14580948 -0.05015092
 -0.59886813 -0.01454399  0.55447122 -0.67386983  0.34830609 -0.69616151
  0.00856163  0.09599129 -0.50840394  0.0845784  -0.40932511 -0.05446706
 -0.74440592 -0.78302727 -0.22730609 -0.18885883 -1.09468843  0.77302618
  0.09184274 -0.61724418 -0.1576483   0.32280642 -0.04405735 -0.69858147
 -0.51081936  0.10842524  0.95211691  0.0071222  -1.00626418 -1.09134218
 -0.41260983 -0.44863226 -0.81963786  0.01921836  0.45686051  0.59322104
  0.59589368 -0.27178954  0.11282168  0.11282168 -0.99624553  0.6602568
  0.02693892  0.45096323  0.37325641 -0.02779038  0.13262084]
w2: [-0.04841506 -1.54246462 -0.22043579 -0.84623993 -0.25187253  1.05957999
  0.13577786 -0.08677695 -0.24954619  0.41355016 -0.59747757  0.78475009
 -0.08509388  0.          0.        

### Question 2: NN with SVC Loss

Make a new neural network fitting function, but with SVC loss instead. You only need to implement this with ReLu activation for the first layer, and Sigmoid for the second.

*(Hint: remember that we need to represent the y-values as 1 or -1 to use this loss function.)*

Use your function to fit an Indica vs Sativa model on a 80% training set from the cannabis data.

In [37]:
def relu(x):
    return np.maximum(0, x)

# hinge loss (SVC loss)
def hinge_loss(pred, target):
    return np.maximum(0, 1 - target * pred)

# derivative of hinge loss with respect to the prediction
def hinge_loss_derivative(pred, target):
    return-target if pred < 1 else 0

def check_stopping_condition(v, d, threshold=1e-4):
    return np.linalg.norm(v) < threshold and np.abs(d) < threshold

In [38]:
def fit_perceptron_svc(train_pred, train_target, w1, w2, w3, c1, c2, c3, learning_rate=0.01, epochs=1000):
    # Initialize v and d
    v = np.array([1/3, 1/3, 1/3])
    d = 0
    
    # convert target to -1 and 1
    train_target = np.where(train_target == 0, -1, 1)

    for epoch in range(1, epochs+1):
        total_loss = 0
        
        for preds, target in zip(train_pred, train_target):
            preds = np.array(preds)
            z1 = np.dot(preds, w1) + c1
            z2 = np.dot(preds, w2) + c2
            z3 = np.dot(preds, w3) + c3
            u1 = relu(z1)
            u2 = relu(z2)
            u3 = relu(z3)
            u = np.array([u1, u2, u3])

            # hidden layer
            hidden_layer_input = np.dot(u, v) + d
            output = sigmoid(hidden_layer_input)
            
            # compute loss
            current_loss = hinge_loss(output, target)
            total_loss += current_loss
            
            # backpropagation
            output_error = hinge_loss_derivative(output, target)
            deriv = sigmoid_der(output)
            
            # gradients for v and d
            grad_v = output_error * deriv * u
            grad_d = output_error * deriv
            
            # gradients for w's and c's
            grad_u = output_error * deriv * v
            grad_z1 = grad_u[0] * (z1 > 0)
            grad_z2 = grad_u[1] * (z2 > 0)
            grad_z3 = grad_u[2] * (z3 > 0)
            grad_w1 = grad_z1 * preds
            grad_w2 = grad_z2 * preds
            grad_w3 = grad_z3 * preds
            grad_c1 = grad_z1
            grad_c2 = grad_z2
            grad_c3 = grad_z3
            
            # update params
            w1 -= learning_rate * grad_w1
            w2 -= learning_rate * grad_w2
            w3 -= learning_rate * grad_w3
            c1 -= learning_rate * grad_c1
            c2 -= learning_rate * grad_c2
            c3 -= learning_rate * grad_c3
            v -= learning_rate * grad_v
            d -= learning_rate * grad_d

        if check_stopping_condition(v, d, 1e-4):
            break

    return w1, w2, w3, c1, c2, c3, v, d

In [39]:
# ignore warnings
import warnings
warnings.filterwarnings('ignore')
def predict_svc(preds, w1, w2, w3, c1, c2, c3, v, d):
    preds = np.array(preds)
    z1 = np.dot(preds, w1) + c1
    z2 = np.dot(preds, w2) + c2
    z3 = np.dot(preds, w3) + c3
    u1 = relu(z1)
    u2 = relu(z2)
    u3 = relu(z3)
    u = np.array([u1, u2, u3])
    # Hidden layer output
    hidden_layer_input = np.dot(u, v) + d
    output = sigmoid(hidden_layer_input)

    return output * 100

svc_w1, svc_w2, svc_w3, svc_c1, svc_c2, svc_c3, svc_v, svc_d = fit_perceptron_svc(X_train, y_train, w1, w2, w3, c1, c2, c3)

# print values
print('SVC')
print('w1:', svc_w1)
print('w2:', svc_w2)
print('w3:', svc_w3)
print('c1:', svc_c1)
print('c2:', svc_c2)
print('c3:', svc_c3)
print('v:', svc_v)
print('d:', svc_d)

SVC
w1: [ 1.74907501e+00 -9.84223499e+00 -3.61494796e+01 -1.21407108e+00
 -3.10341934e+01 -3.24834144e+01 -3.93728573e-01  1.09853916e+01
 -5.24278386e+00  1.75691949e+01 -1.39809669e+01  2.88459984e+01
 -4.11026028e+01  0.00000000e+00  0.00000000e+00  4.49056398e+00
  2.30507834e+01  2.94350882e+01 -1.10735864e+00 -1.56819797e-02
  3.43280771e+01 -8.27868388e+00  2.40920096e-01 -1.81014108e+01
  7.49706117e-02  9.07279533e+00  7.62553980e+00 -7.22878567e+00
 -2.36213700e+01  2.14067039e+01  4.88739285e+00 -1.10841861e+01
 -7.12001212e+00 -2.22544112e+00 -1.24756225e+01  7.21840982e-01
  3.17679381e-01 -8.24662967e-01  4.08685870e+00  5.85358956e-01
 -4.22253044e-02 -3.35964245e+00 -1.15396746e+00  1.66269341e+00
  2.07392698e+00 -7.08799301e+00  7.81826734e+00 -1.24722762e+01
 -7.00857351e+00  7.94785908e+00 -8.19637857e-01 -5.01335387e-01
  6.59929770e+00  1.28767226e+01  2.90600998e+00 -3.52178610e+00
  9.41207311e+00  9.41207311e+00 -1.64497882e+00  8.88650080e+00
  2.69389195e-02 

### Question 3: Comparison

Which of the four Neural Networks (squared loss sigmoid, relu, or softplus; and SVC loss) performed the best on the remaining 20% test set?

In [40]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# sigmoid activation
y_pred = [predict_with_fitted_model(pred, sig_w1, sig_w2, sig_w3, sig_c1, sig_c2, sig_c3, sig_v, sig_d) for pred in X_test]
y_pred = [1 if pred > 50 else 0 for pred in y_pred]
print('Sigmoid Activation')
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('F1 Score:', f1_score(y_test, y_pred))
print()

# relu activation
y_pred = [predict_with_fitted_model(pred, relu_w1, relu_w2, relu_w3, relu_c1, relu_c2, relu_c3, relu_v, relu_d, activation='relu') for pred in X_test]
y_pred = [1 if pred > 50 else 0 for pred in y_pred]
print('ReLU Activation')
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('F1 Score:', f1_score(y_test, y_pred))
print()

# softplus activation
y_pred = [predict_with_fitted_model(pred, softplus_w1, softplus_w2, softplus_w3, softplus_c1, softplus_c2, softplus_c3, softplus_v, softplus_d, activation='softplus') for pred in X_test]
y_pred = [1 if pred > 50 else 0 for pred in y_pred]
print('Softplus Activation')
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('F1 Score:', f1_score(y_test, y_pred))
print()

# svc
y_pred = [predict_svc(pred, svc_w1, svc_w2, svc_w3, svc_c1, svc_c2, svc_c3, svc_v, svc_d) for pred in X_test]
y_pred = [1 if pred > 50 else 0 for pred in y_pred]
print('SVC')
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('F1 Score:', f1_score(y_test, y_pred))

Sigmoid Activation
Accuracy: 0.7354260089686099
Precision: 0.6341463414634146
Recall: 0.37142857142857144
F1 Score: 0.46846846846846846

ReLU Activation
Accuracy: 0.8295964125560538
Precision: 0.6951219512195121
Recall: 0.8142857142857143
F1 Score: 0.75

Softplus Activation
Accuracy: 0.8340807174887892
Precision: 0.7037037037037037
Recall: 0.8142857142857143
F1 Score: 0.7549668874172185

SVC
Accuracy: 0.852017937219731
Precision: 0.7761194029850746
Recall: 0.7428571428571429
F1 Score: 0.759124087591241


The model with the sigmoid activation function and SVC loss performed the best on the test set with an accuracy of 0.85 and an F1 score of 0.76. The models with the ReLU and Softplus activation functions had better accuracy scores however their F1 scores were just on par with the SVC loss model. THe model that did the worst was the model with the squared loss and sigmoid activation function.

## Part C: Concepts

### Question 1: interpret the weights

Investigate the fitted weights in the first layer of one of your networks from B1 and B2: First look at the scores, $z_1, z_2, z_3$ and see if they correlate to anything related to the data. If they do - for example, if $z_1$ seems to capture a meaningful data measurement - interpret some of the corresponding weights.

- Taking a look at the weights for the first layer of the model with the sigmoid activation function and squared loss, there does not seem to be any distinct relationship with the data. However the higher z values do indicate more influential features for the final predictions. But from the weights there is no single feature that stands out as being more important than the others. The variation in the z values means that the relationship is more complex that just relying on a single feature.

### Question 2: Multiclass model

Suppose you had been asked to instead implement an NN with three outputs: P(Indica), P(Sativa), P(Hybrid).

Explain what parts of you code would need to change, and what parts could stay the same.

- Parts that need to change:
    - Output layer would need to have 3 nodes instead of 1 since we are now returning 3 values
    - Also the activation function would need to changed to a softmax function since we are now dealing with multiple classes. This would also need to be normalized to sum to 1 as they are probabilities.
    - The loss function would also need to be adjusted since we are now dealing with multiple classes. The loss function would need to be a cross-entropy loss function. This works well for multi-class classification problems and is the negative log likelihood of the true labels given the predicted probabilities.
    - Finally the target variable would need to be one-hot encoded to represent the three classes. So instead of binary values we could have an array of 1 or 0 to indicate which class the observation belongs to.
- Parts that could stay the same:
    - General neural network structure would remain the same since we would still have the input, then hidden layer, and then output layer. There would be the same number of layers, but just with more nodes in the output layer.
    - The weights would still be the same to initialize the model. Also the input and hidden layers would stay the same since they are not dependent on the number of classes.